File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int solve(string s, string t, int x, int y)
    {
        if(x == 0 && y == 0) return 1;
        if(x == 0 || y == 0) return 0;
        
        int ans = 0;
        if(s[x-1] == t[y-1])
            ans += solve(s, t, x-1, y-1);
            
        ans += solve(s, t, x-1, y);
        return ans;
    }
    
    int numDistinct(string s, string t) {
        int x = s.length(), y = t.length();
        
        vector<vector<int>> dp(x+1, vector<int> (y+1, 0));
        for(int i = 0; i <= x; i++)
        {
            for(int j = 0; j <= y; j++)
            {
                long long ans = 0;
                if(j == 0) {
                    dp[i][j] = 1;
                    continue;
                }
                if(i == 0) {
                    dp[i][j] = 0;
                    continue;
                }
                
                if(s[i-1] == t[j-1])
                    ans += (long long) dp[i-1][j-1];
                
                ans += (long long) dp[i-1][j];
                
                dp[i][j] = ans;
            }
        }
        
        return (int) dp[x][y];
    }
};

/*

babgbag
bag





*/

File: 0115-distinct-subsequences/0115-distinct-subsequences.java
class Solution {
    int[][] dp;
    public int lcs(String s, String t, int n, int m)
    {
        if(m == 0) return 1;
        if(n == 0 || m == 0) return 0;

        if(dp[n][m] != -1) return dp[n][m];

        int x = 0;
        if(s.charAt(n-1) == t.charAt(m-1)) {
            x = lcs(s, t, n-1, m-1);
        }

        return dp[n][m] = x + lcs(s, t, n-1, m);
        
    }
    public int numDistinct(String s, String t) 
    {
        int n = s.length(), m = t.length();
        dp = new int[n+1][m+1];
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= m; j++) {
                dp[i][j] = -1;
            }
        }
        return lcs(s, t, n, m);    
    }
}